src/cookbooks/client_guidelines.md

```diff
@@ -0,0 +1,137 @@
+# Client Implementation Guidelines for Experiments
+
+There are three supported approaches to developing experimental features for Firefox:
+
+- [Feature Gates](#feature-gates)
+  - The preferred method for testing experimental features is to __land in-tree__ and ship behind a __feature gate__.
+- [Firefox Prefs](#prefs)
+  - If feature gates are [not appropriate](#feature-gate-caveats) for your project, Firefox Preferences (AKA "__Prefs__") should be used.
+- [Firefox Extensions](#extensions) AKA "__Add-ons__".
+  - If the feature being tested should not land in the tree, or if it will ultimately ship as an extension, then an extension should be used.
+
+New features go through the standard Firefox review, testing, and deployment processes, and are then enabled experimentally in the field using [Normandy](https://github.com/mozilla/normandy).
+
+## Feature Gates
+
+A new Feature Gate library for Firefox Desktop is now available.
+
+__Each feature gate should represent a different experimental treatment__.
+
+### Feature Gate caveats
+
+The current Feature Gate library comes with a few caveats, and may not be appropriate for your situation:
+
+- Only JS is supported.
+- Always asynchronous.
+
+Future versions of the Feature Gate API will include C++/Rust support and a synchronous API.
+
+### Using the Feature Gate library
+
+Read [the documentation][feature-gate-docs] to get started.
+
+## Prefs
+
+Firefox Preferences (AKA "prefs") are commonly used to enable and disable features. However, prefs are more complex to implement correctly than [feature gates](#feature-gates).
+
+__Each pref should represent a different experimental treatment__.
+
+There are three types of Prefs:
+
+1. Built-in prefs - shipped with Firefox, in `firefox.js`.
+2. `user branch` - set by the user, overriding built-in prefs.
+3. `default branch` - Overrides both built-in and `user branch` prefs. Only persists until the browser session ends, next restart will revert to either built-in or `user branch` (if set).
+
+  [Normandy](https://github.com/mozilla/normandy) supports overriding both the `user` and `default` branches, although the latter is preferred as it does not permanently override user settings. `default` branch prefs are simple to reset since they do not persist past a restart.
+
+   __In order for features to be activated experimentally, these rules must be observed__:
+
+- The feature must not start up before `final-ui-startup` is observed.
```

Comment:
Sure, I think it's basically `Services.obs.addObserver(this, "final-ui-startup", true);1`... I am wary of putting too many code samples and links to code etc. since things can change so much over time and prefer to link to docs... however I think we should go ahead and do this because prefs are going to yield to feature gates completely, eventually, and prefs are unlikely to change in the meantime that much (I hope!)

Reply:
I suppose we could link to https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Reference/Interface/nsIObserverService#addObserver() but it's pretty confusing, because `Services.obs` is just a convenient interface to this service.

Reply:
Hmm... yeah that could work or this page on ][Observer Notifications](https://developer.mozilla.org/en-US/docs/Mozilla/Tech/XPCOM/Observer_Notifications) which references `final-ui-startup` explicitly?